grayscale.py

```diff
@@ -1,764 +1,549 @@
+        dBuf = ptr32(self._drawBuffer)
+        b1 = ptr32(buffers[0]); b2 = ptr32(buffers[1]); b3 = ptr32(buffers[2])
+
+        dc(0) # command mode
+        state[_ST_THREAD] = _THREAD_RUNNING
+        while state[_ST_THREAD] == _THREAD_RUNNING:
+            # This is the main GPU loop.
+            # We cycle through each of the 3 display subframe buffers,
+            # sending the framebuffer data and various commands.
+            for fn in range(3):
+                # Send fn'th subframe
+                t0 = ticks_us()
+                # Send the pre-frame commands to 'park' the row counter
+                spi_write(preFrameCmds)
+                # Now send the (sub) frame data
+                dc(1) # frame data mode
+                spi_write(buffers[fn])
+                dc(0) # back to command mode
+                # Send the first instance of the contrast adjust command
+                spi_write(postFrameAdj[fn])
+                # Wait for the pre-frame time to complete
+                sleep_us(_PRE_FRAME_TIME - int(ticks_diff(ticks_us(), t0)))
+                t0 = ticks_us()
+                # Send the post-frame commands to display the frame
+                spi_write(postFrameCmds)
+                # Adjust the contrast for the specific frame number again.
+                # If we do not do this twice, the screen can glitch.
+                spi_write(postFrameAdj[fn])
+
+                # Tasks that only happen after the last subframe
+                if (fn == 2):
+                    # Check if there's a pending frame copy required.
+                    # We only copy the paint framebuffers to the display
+                    # framebuffers on the last frame to avoid screen-tearing.
+                    if (state[_ST_COPY_BUFFS] != 0):
+                        # By using using ptr32 vars we copy 4 bytes at a time
+                        for i in range (_BUFF_INT_SIZE):
+                            v1 = dBuf[i]
+                            v2 = dBuf[i+_BUFF_INT_SIZE]
+                            # This remaps to the different buffer format.
+                            b1[i] = v1 | v2 # white, lightgray and darkgray
+                            b2[i] = v1 # white and lightgray
+                            b3[i] = v1 & (v1^v2) # white only
+                        state[_ST_COPY_BUFFS] = 0
+                    # Check if there's a pending contrast/brightness change
+                    elif (state[_ST_CONTRAST] != 0xff):
+                        contrast = state[_ST_CONTRAST]
+                        state[_ST_CONTRAST] = 0xff
+                        # Shift the value to provide 3 different levels
+                        postFrameAdj[0][1] = contrast >> 5
+                        postFrameAdj[1][1] = contrast >> 1
+                        postFrameAdj[2][1] = (contrast << 1) + 1
+                    # Check if there are pending display controller commands
+                    elif state[_ST_PENDING_CMD]:
```

Comment:
Yeah, I noticed that. I didn't think it needed that level of responsiveness. We only use it for display power on and off so far. Better to keep the GPU tighter than checking it all each subframe, no?

Reply:
But I hear you, out of scope.

Reply:
That's a fair point about the latency. And it could be argued that switching the display on and off should only be done at the end of subframe 2.
It probably doesn't make any difference, but actually performing the same set of checks on each subframe might be preferable to only doing these checks on subframe 2, as that would mean that each subframe iteration is closer to being the same amount of opcodes executed. But like I said that difference is probably negligible, and locking the commands to subframe 2 end (aka subframe 0 start) is probably better.